File: internal/manager/agent/cleanup.go
package agent

import (
	"bytes"
	"context"
	"crypto/subtle"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"

	"github.com/codeasier/mtls-router/internal/manager/agent/modelconfig"
)

type CleanupPreviewRequest struct {
	Agent Kind
}

type CleanupWriteRequest struct {
	Agent                   Kind
	RevisionToken           string
	ApproveManagedOverwrite bool
}

type CleanupPreview struct {
	RevisionToken      string        `json:"revision_token"`
	Agent              Kind          `json:"agent"`
	Files              []FilePreview `json:"files"`
	RemovedPaths       []string      `json:"removed_paths"`
	ManagedConfigDrift bool          `json:"managed_config_drift"`
	StateChange        *FilePreview  `json:"state_change,omitempty"`
	StateBackup        *FilePreview  `json:"state_backup,omitempty"`
}

type cleanupPlan struct {
	writePlan
	removedByPath  map[string][]string
	stateOperation Operation
}

// CleanupPreview builds a key-free cleanup plan from authenticated last-applied
// state. It does not derive paths from the current environment or create state.
func (s *Service) CleanupPreview(ctx context.Context, request CleanupPreviewRequest) (CleanupPreview, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := contextError(ctx); err != nil {
		return CleanupPreview{}, err
	}
	if err := validateCleanupAgent(request.Agent); err != nil {
		return CleanupPreview{}, err
	}
	if err := s.ensureCleanupStateExists(request.Agent); err != nil {
		return CleanupPreview{}, err
	}
	lock, err := acquireExistingTransactionLock(ctx, s.stateDir)
	if err != nil {
		if CodeOf(err) != CodeOperationBusy {
			return CleanupPreview{}, operationError(CodeModelStateInvalid, "Agent cleanup coordination state is invalid")
		}
		return CleanupPreview{}, err
	}
	defer lock.release()
	if err := contextError(ctx); err != nil {
		return CleanupPreview{}, err
	}
	if err := s.ensureCleanupStateExists(request.Agent); err != nil {
		return CleanupPreview{}, err
	}
	if s.writesDisabled {
		return CleanupPreview{}, operationError(CodeRollbackFailed, "Agent writes are disabled because transaction recovery failed")
	}
	if err := s.ensureExistingSigner(); err != nil {
		return CleanupPreview{}, err
	}
	plan, err := s.buildCleanupPlan(request.Agent)
	if err != nil {
		return CleanupPreview{}, err
	}
	return s.cleanupPreviewForPlan(plan)
}

// CleanupWrite re-creates and verifies the signed cleanup plan before routing
// it through the normal backup, journal, apply, and rollback machinery.
func (s *Service) CleanupWrite(ctx context.Context, request CleanupWriteRequest) (WriteResult, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := contextError(ctx); err != nil {
		return WriteResult{}, err
	}
	if err := validateCleanupAgent(request.Agent); err != nil {
		return WriteResult{}, err
	}
	if err := s.ensureCleanupStateExists(request.Agent); err != nil {
		return WriteResult{}, err
	}
	lock, err := acquireTransactionLock(ctx, s.stateDir)
	if err != nil {
		return WriteResult{}, err
	}
	defer lock.release()
	if err := contextError(ctx); err != nil {
		return WriteResult{}, err
	}
	if err := s.ensureCleanupStateExists(request.Agent); err != nil {
		return WriteResult{}, err
	}
	if s.writesDisabled {
		return WriteResult{}, operationError(CodeRollbackFailed, "Agent writes are disabled because transaction recovery failed")
	}
	if request.RevisionToken == "" {
		return WriteResult{}, operationError(CodeInvalidParams, "cleanup revision token is required")
	}
	if err := s.ensureExistingSigner(); err != nil {
		return WriteResult{}, err
	}
	claims, err := s.signer.VerifyCleanupRevision(request.RevisionToken)
	if err != nil || claims.Agent != modelAgent(request.Agent) {
		return WriteResult{}, operationError(CodePreviewStale, "Agent cleanup preview is stale")
	}
	if err := s.verifyKeyedRevision(s.sidecarPath(), fileRevisionFromClaim(claims.StateRevision), revisionContextSidecar); err != nil {
		return WriteResult{}, operationError(CodePreviewStale, "Agent cleanup state changed after preview")
	}
	for _, file := range claims.Files {
		if err := s.verifyKeyedRevision(file.SourcePath, fileRevisionFromClaim(file.SourceRevision), revisionContextAgentFile); err != nil {
			return WriteResult{}, operationError(CodePreviewStale, "Agent configuration changed after cleanup preview")
		}
	}
	plan, err := s.buildCleanupPlan(request.Agent)
	if err != nil {
		return WriteResult{}, err
	}
	currentToken, err := s.cleanupTokenForPlan(plan)
	if err != nil || subtle.ConstantTimeCompare([]byte(currentToken), []byte(request.RevisionToken)) != 1 {
		return WriteResult{}, operationError(CodePreviewStale, "Agent cleanup preview is stale")
	}
	if plan.ManagedConfigDrift() != request.ApproveManagedOverwrite {
		return WriteResult{}, operationError(CodeManagedConfigDrift, "managed Agent configuration drift approval does not match preview")
	}
	if err := contextError(ctx); err != nil {
		return WriteResult{}, err
	}
	return s.executePlan(ctx, plan.writePlan, nil)
}

func validateCleanupAgent(kind Kind) error {
	switch kind {
	case ClaudeCode, OpenCode, Codex:
		return nil
	default:
		return operationError(CodeInvalidParams, "unsupported Agent cleanup selection")
	}
}

func (s *Service) ensureCleanupStateExists(kind Kind) error {
	if _, err := os.Stat(s.sidecarPath()); os.IsNotExist(err) {
		return operationError(CodeAgentNotManaged, agentName(kind)+" is not managed by CodeasierRouter")
	} else if err != nil {
		return operationError(CodeModelStateInvalid, "Agent model state cannot be read")
	}
	return nil
}

func (s *Service) buildCleanupPlan(kind Kind) (cleanupPlan, error) {
	state, stateRevision, stateContent, stateMode, err := s.readSidecar()
	if err != nil {
		return cleanupPlan{}, err
	}
	section, managed := state.Agents[kind]
	if !managed {
		return cleanupPlan{}, operationError(CodeAgentNotManaged, agentName(kind)+" is not managed by CodeasierRouter")
	}
	saved, err := decodeCleanupModelConfig(kind, section.ModelConfig)
	if err != nil {
		return cleanupPlan{}, operationError(CodeModelStateInvalid, "saved Agent model state is invalid")
	}
	plan := cleanupPlan{
		writePlan: writePlan{
			agents: []agentPlan{{kind: kind, mode: ConfigModeMerge}}, selected: []Kind{kind},
			sidecar: state, sidecarRevision: stateRevision, sidecarContent: stateContent, sidecarMode: stateMode,
		},
		removedByPath: make(map[string][]string),
	}
	for _, recorded := range section.Files {
		file, removed, drifted, err := s.cleanupFilePlan(kind, recorded, section, saved)
		if err != nil {
			return cleanupPlan{}, err
		}
		if drifted {
			plan.drifted = []Kind{kind}
		}
		if len(removed) == 0 {
			continue
		}
		plan.files = append(plan.files, file)
		plan.removedByPath[file.targetPath] = removed
	}
	delete(state.Agents, kind)
	plan.sidecar = state
	if len(state.Agents) == 0 {
		plan.stateOperation = OperationDelete
		plan.files = append(plan.files, plannedFile{
			scope: scopeManagerState, role: "state", format: FormatJSON,
			sourcePath: s.sidecarPath(), targetPath: s.sidecarPath(), operation: OperationDelete,
			sourceRevision: stateRevision, targetRevision: stateRevision, sourceContent: stateContent,
			sourceMode: stateMode, targetMode: stateMode, backupRequired: true,
			backupSource: s.sidecarPath(), restoreFrom: s.sidecarPath(),
		})
		return plan, nil
	}
	stateOutput, err := modelconfig.CanonicalValue(state)
	if err != nil || len(stateOutput) > maxSidecarSize {
		return cleanupPlan{}, operationError(CodeModelStateInvalid, "Agent model state could not be updated")
	}
	plan.stateOperation = OperationReplace
	plan.files = append(plan.files, plannedFile{
		scope: scopeManagerState, role: "state", format: FormatJSON,
		sourcePath: s.sidecarPath(), targetPath: s.sidecarPath(), operation: OperationReplace,
		sourceRevision: stateRevision, targetRevision: stateRevision, sourceContent: stateContent,
		sourceMode: stateMode, targetMode: stateMode, backupRequired: true,
		backupSource: s.sidecarPath(), restoreFrom: s.sidecarPath(),
		render: func([]byte) ([]byte, error) { return append([]byte(nil), stateOutput...), nil },
	})
	return plan, nil
}

func decodeCleanupModelConfig(kind Kind, raw json.RawMessage) (*modelconfig.Config, error) {
	document, err := modelconfig.CanonicalValue(map[string]any{
		"version":                modelconfig.Version,
		string(modelAgent(kind)): json.RawMessage(raw),
	})
	if err != nil {
		return nil, err
	}
	return modelconfig.DecodeStructural(document)
}

func (s *Service) cleanupFilePlan(kind Kind, recorded lastAppliedFile, section lastAppliedAgent, saved *modelconfig.Config) (plannedFile, []string, bool, error) {
	safety := inspectRecoveryTarget(recorded.Role, recorded.Path, cleanupFileFormat(kind, recorded))
	if len(safety.Reasons) != 0 {
		if containsRecoveryReason(safety.Reasons, RecoveryNotWritable) || containsRecoveryReason(safety.Reasons, RecoveryParentUnavailable) {
			return plannedFile{}, nil, false, operationError(CodeConfigNotWritable, agentName(kind)+" managed configuration is not writable")
		}
		return plannedFile{}, nil, false, operationError(CodeConfigInvalid, agentName(kind)+" managed configuration path is unsafe")
	}
	revision, content, mode, err := s.readKeyedRevision(recorded.Path, revisionContextAgentFile)
	if err != nil || !revision.Exists {
		return plannedFile{}, nil, false, operationError(CodeConfigInvalid, agentName(kind)+" managed configuration cannot be read")
	}
	if err := ensureWritable(recorded.Path); err != nil {
		return plannedFile{}, nil, false, err
	}
	var transform cleanupTransform
	format := FormatJSON
	switch kind {
	case ClaudeCode:
		root, valid := decodeObject(content)
		if !valid {
			return plannedFile{}, nil, false, operationError(CodeConfigInvalid, "Claude Code configuration is not a JSON object")
		}
		transform, err = cleanupClaude(root, section.OwnedPaths)
	case OpenCode:
		parsed := content
		if filepath.Ext(recorded.Path) == ".jsonc" {
			format = FormatJSONC
			parsed, err = stripJSONC(content)
			if err != nil {
				return plannedFile{}, nil, false, operationError(CodeConfigInvalid, "opencode configuration is invalid JSONC")
			}
		}
		root, valid := decodeObject(parsed)
		if !valid {
			return plannedFile{}, nil, false, operationError(CodeConfigInvalid, "opencode configuration is not a JSON object")
		}
		transform, err = cleanupOpenCode(root, containsString(section.OwnedPaths, "model"))
	case Codex:
		if recorded.Role == "config" {
			format = FormatTOML
			transform, err = cleanupCodexConfig(content, saved.Codex)
		} else {
			root, valid := decodeObject(content)
			if !valid {
				return plannedFile{}, nil, false, operationError(CodeConfigInvalid, "Codex auth configuration is not a JSON object")
			}
			transform, err = cleanupCodexAuth(root)
		}
	}
	if err != nil {
		return plannedFile{}, nil, false, err
	}
	operation := OperationReplace
	if transform.Delete {
		operation = OperationDelete
	}
	output := append([]byte(nil), transform.Content...)
	file := plannedFile{
		agent: kind, mode: ConfigModeMerge, role: recorded.Role, format: format, scope: scopeAgent,
		sourcePath: recorded.Path, targetPath: recorded.Path, operation: operation,
		containsAPIKey: kind != Codex || recorded.Role == "auth",
		sourceRevision: revision, targetRevision: revision, sourceContent: content,
		sourceMode: mode, targetMode: mode, backupRequired: true,
		backupSource: recorded.Path, restoreFrom: recorded.Path,
	}
	if operation == OperationReplace {
		file.render = func([]byte) ([]byte, error) { return append([]byte(nil), output...), nil }
	}
	return file, transform.RemovedPaths, revisionTokenValue(revision) != recorded.RevisionMAC, nil
}

func cleanupFileFormat(kind Kind, recorded lastAppliedFile) Format {
	if kind == Codex && recorded.Role == "config" {
		return FormatTOML
	}
	if kind == OpenCode && filepath.Ext(recorded.Path) == ".jsonc" {
		return FormatJSONC
	}
	return FormatJSON
}

func (p cleanupPlan) ManagedConfigDrift() bool { return len(p.drifted) != 0 }

func (s *Service) cleanupTokenForPlan(plan cleanupPlan) (string, error) {
	files := make([]modelconfig.CleanupRevisionFile, 0, len(plan.files)-1)
	for _, file := range plan.files {
		if file.scope == scopeManagerState {
			continue
		}
		files = append(files, modelconfig.CleanupRevisionFile{
			Role: file.role, SourcePath: filepath.Clean(file.sourcePath), TargetPath: filepath.Clean(file.targetPath),
			Operation: string(file.operation), BackupRequired: file.backupRequired, BackupSource: file.backupSource,
			SourceRevision: revisionClaim(file.sourceRevision), TargetRevision: revisionClaim(file.targetRevision),
			RemovedPaths: append([]string(nil), plan.removedByPath[file.targetPath]...),
		})
	}
	return s.signer.SignCleanupRevision(modelconfig.CleanupRevisionClaims{
		Agent: modelAgent(plan.selected[0]), Files: files, StateOperation: string(plan.stateOperation),
		StateRevision: revisionClaim(plan.sidecarRevision), ManagedConfigDrift: plan.ManagedConfigDrift(),
	})
}

func fileRevisionFromClaim(value modelconfig.RevisionState) fileRevision {
	return fileRevision{Exists: value.Exists, Size: value.Size, Mode: value.Mode, Digest: value.Digest}
}

func (s *Service) cleanupPreviewForPlan(plan cleanupPlan) (CleanupPreview, error) {
	token, err := s.cleanupTokenForPlan(plan)
	if err != nil {
		return CleanupPreview{}, operationError(CodeModelStateInvalid, "Agent cleanup revision token could not be created")
	}
	preview := CleanupPreview{
		RevisionToken: token, Agent: plan.selected[0], ManagedConfigDrift: plan.ManagedConfigDrift(),
		StateChange: &FilePreview{
			Role: "state", Path: s.sidecarPath(), Format: FormatJSON, Operation: plan.stateOperation,
			Operations: []Operation{plan.stateOperation}, Backup: cleanupBackupPlan(s.sidecarPath()),
		},
		StateBackup: &FilePreview{
			Role: "state", Path: s.sidecarPath(), Format: FormatJSON, Operation: Operation("backup"),
			Operations: []Operation{Operation("backup")}, Backup: cleanupBackupPlan(s.sidecarPath()),
		},
	}
	for _, file := range plan.files {
		if file.scope == scopeManagerState {
			continue
		}
		removed := plan.removedByPath[file.targetPath]
		preview.RemovedPaths = append(preview.RemovedPaths, removed...)
		preview.Files = append(preview.Files, FilePreview{
			Role: file.role, Path: file.targetPath, Format: file.format, Operation: file.operation,
			Operations: []Operation{file.operation}, ContainsAPIKey: file.containsAPIKey,
			Backup: cleanupBackupPlan(file.backupSource),
		})
	}
	sort.Strings(preview.RemovedPaths)
	preview.RemovedPaths = uniqueSortedStrings(preview.RemovedPaths)
	return preview, nil
}

func cleanupBackupPlan(path string) BackupPlan {
	return BackupPlan{
		Required: true, Sensitive: true, Pattern: path + ".bak-<timestamp>-<random>",
	}
}

func uniqueSortedStrings(values []string) []string {
	if len(values) == 0 {
		return values
	}
	result := values[:1]
	for _, value := range values[1:] {
		if value != result[len(result)-1] {
			result = append(result, value)
		}
	}
	return result
}

type cleanupTransform struct {
	Content      []byte
	Delete       bool
	RemovedPaths []string
}

func newCleanupTransform(content []byte, empty bool, removed []string) cleanupTransform {
	sort.Strings(removed)
	unique := removed[:0]
	for _, path := range removed {
		if len(unique) == 0 || unique[len(unique)-1] != path {
			unique = append(unique, path)
		}
	}
	if empty {
		content = nil
	}
	return cleanupTransform{Content: content, Delete: empty, RemovedPaths: unique}
}

func marshalCleanupJSON(value any) ([]byte, error) {
	var output bytes.Buffer
	encoder := json.NewEncoder(&output)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(value); err != nil {
		return nil, err
	}
	return output.Bytes(), nil
}

File: desktop/src-tauri/src/manager_core/agent/cleanup.rs
use std::path::Path;

use super::detect::{decode_object, strip_jsonc};
use super::lock::acquire_existing_transaction_lock;
use super::modelconfig::{
    canonical_value, decode_structural, Agent, CleanupRevisionClaims, Config, JsonNumber,
    JsonValue, TokenSigner,
};
use super::plan::{
    agent_name, cleanup_revision_file, AgentPlan, BackupPlan, ConfigMode, FilePreview, FileRender,
    FileRevision, JournalScope, Operation, PlannedFile, PlannedRevisionGuard, WritePlan,
};
use super::recovery::{contains_recovery_reason, inspect_recovery_target};
use super::render::{cleanup_claude, cleanup_codex_auth, cleanup_codex_config, cleanup_opencode};
use super::revision::read_keyed_revision;
use super::sidecar::{
    read_sidecar, sidecar_path, LastAppliedAgent, LastAppliedFile, LastAppliedState,
};
use super::types::{Format, OperationError, RecoveryReason};
use crate::protocol::ErrorCode;

#[derive(Clone, Debug)]
pub struct CleanupPreview {
    pub revision_token: String,
    pub agent: Agent,
    pub files: Vec<FilePreview>,
    pub removed_paths: Vec<String>,
    pub managed_config_drift: bool,
    pub state_change: Option<FilePreview>,
    pub state_backup: Option<FilePreview>,
}

pub struct CleanupPlan {
    pub write: WritePlan,
    pub removed_by_path: std::collections::HashMap<String, Vec<String>>,
    pub token_files: Vec<super::modelconfig::CleanupRevisionFile>,
    pub state_operation: Operation,
}

impl CleanupPlan {
    pub fn managed_config_drift(&self) -> bool {
        !self.write.drifted.is_empty()
    }
}

pub fn build_cleanup_plan(
    signer: &TokenSigner,
    state_dir: &Path,
    generation: &str,
    kind: Agent,
) -> Result<CleanupPlan, OperationError> {
    let (mut state, state_revision, state_content, state_mode) =
        read_sidecar(signer, state_dir, generation)?;
    let section = state.agents.get(&kind).cloned().ok_or_else(|| {
        OperationError::new(
            ErrorCode::AgentNotManaged,
            format!("{} is not managed by CodeasierRouter", agent_name(kind)),
        )
    })?;
    let saved = decode_cleanup_model_config(kind, &section.model_config)?;
    let mut plan = CleanupPlan {
        write: WritePlan {
            agents: vec![AgentPlan {
                kind,
                mode: ConfigMode::Merge,
            }],
            selected: vec![kind],
            files: Vec::new(),
            revision_guards: Vec::new(),
            input: Default::default(),
            catalog: super::modelconfig::CatalogClaims {
                version: 1,
                models: vec!["unused".into()],
                agents: vec![kind],
                owner: "desktop".into(),
                router_base_url: "http://127.0.0.1:1".into(),
                deployment_id: "cleanup".into(),
                protocol_version: crate::protocol::MANAGEMENT_PROTOCOL_VERSION.into(),
                simplify: false,
                canonicalization: String::new(),
                key_generation: String::new(),
            },
            canonical: b"{}".to_vec(),
            sidecar: state.clone(),
            sidecar_revision: state_revision.clone(),
            sidecar_content: state_content.clone(),
            sidecar_mode: state_mode,
            drifted: Vec::new(),
            collisions: Vec::new(),
            requires_codex_auth_approval: false,
        },
        removed_by_path: std::collections::HashMap::new(),
        token_files: Vec::new(),
        state_operation: Operation::Replace,
    };
    for recorded in &section.files {
        let (file, removed, drifted) = cleanup_file_plan(signer, kind, recorded, &section, &saved)?;
        if drifted {
            plan.write.drifted = vec![kind];
        }
        if !file.source_revision.exists {
            plan.token_files.push(cleanup_revision_file(&file, &[]));
            plan.write.revision_guards.push(PlannedRevisionGuard {
                path: file.source_path.clone(),
                revision: file.source_revision.clone(),
                scope: JournalScope::Agent,
            });
            continue;
        }
        if removed.is_empty() && file.operation != Operation::Delete {
            continue;
        }
        plan.removed_by_path
            .insert(file.target_path.clone(), removed.clone());
        plan.token_files
            .push(cleanup_revision_file(&file, &removed));
        plan.write.files.push(file);
    }
    state.agents.remove(&kind);
    plan.write.sidecar = state.clone();
    let sidecar = sidecar_path(state_dir).to_string_lossy().into_owned();
    if state.agents.is_empty() {
        plan.state_operation = Operation::Delete;
        plan.write.files.push(PlannedFile {
            agent: None,
            mode: ConfigMode::Merge,
            format: Format::Json,
            source_path: sidecar.clone(),
            target_path: sidecar.clone(),
            operation: Operation::Delete,
            contains_api_key: false,
            preserves: Vec::new(),
            warning: String::new(),
            source_revision: state_revision.clone(),
            target_revision: state_revision,
            source_content: state_content,
            source_mode: state_mode,
            target_mode: state_mode,
            backup_required: true,
            backup_source: sidecar.clone(),
            backup_path: String::new(),
            restore_from: sidecar,
            render: None,
            role: "state".into(),
            companion_exists: false,
            syntax_invalid: false,
            scope: JournalScope::ManagerState,
        });
        return Ok(plan);
    }
    let state_output = sidecar_canonical(&state)?;
    plan.state_operation = Operation::Replace;
    plan.write.files.push(PlannedFile {
        agent: None,
        mode: ConfigMode::Merge,
        format: Format::Json,
        source_path: sidecar.clone(),
        target_path: sidecar.clone(),
        operation: Operation::Replace,
        contains_api_key: false,
        preserves: Vec::new(),
        warning: String::new(),
        source_revision: state_revision.clone(),
        target_revision: state_revision,
        source_content: state_content,
        source_mode: state_mode,
        target_mode: state_mode,
        backup_required: true,
        backup_source: sidecar.clone(),
        backup_path: String::new(),
        restore_from: sidecar,
        render: Some(FileRender::Static(state_output)),
        role: "state".into(),
        companion_exists: false,
        syntax_invalid: false,
        scope: JournalScope::ManagerState,
    });
    Ok(plan)
}

fn sidecar_canonical(state: &LastAppliedState) -> Result<Vec<u8>, OperationError> {
    super::sidecar::sidecar_to_canonical(state).and_then(|content| {
        if content.len() > super::sidecar::MAX_SIDECAR_SIZE {
            Err(OperationError::new(
                ErrorCode::ModelStateInvalid,
                "Agent model state could not be updated",
            ))
        } else {
            Ok(content)
        }
    })
}

fn decode_cleanup_model_config(kind: Agent, raw: &JsonValue) -> Result<Config, OperationError> {
    let document = canonical_value(&JsonValue::object([
        ("version".into(), JsonValue::Number(JsonNumber("1".into()))),
        (kind.as_str().to_owned(), raw.clone()),
    ]))
    .map_err(|_| {
        OperationError::new(
            ErrorCode::ModelStateInvalid,
            "saved Agent model state is invalid",
        )
    })?;
    decode_structural(&document).map_err(|_| {
        OperationError::new(
            ErrorCode::ModelStateInvalid,
            "saved Agent model state is invalid",
        )
    })
}

fn cleanup_file_plan(
    signer: &TokenSigner,
    kind: Agent,
    recorded: &LastAppliedFile,
    section: &LastAppliedAgent,
    saved: &Config,
) -> Result<(PlannedFile, Vec<String>, bool), OperationError> {
    let format = cleanup_file_format(kind, recorded);
    let safety = inspect_recovery_target(&recorded.role, Path::new(&recorded.path), format);
    if !safety.exists && !contains_recovery_reason(&safety.reasons, RecoveryReason::Unreadable) {
        return Ok((
            PlannedFile {
                agent: Some(kind),
                mode: ConfigMode::Merge,
                format,
                source_path: recorded.path.clone(),
                target_path: recorded.path.clone(),
                operation: Operation::Delete,
                contains_api_key: false,
                preserves: Vec::new(),
                warning: String::new(),
                source_revision: FileRevision::default(),
                target_revision: FileRevision::default(),
                source_content: Vec::new(),
                source_mode: 0,
                target_mode: 0,
                backup_required: false,
                backup_source: String::new(),
                backup_path: String::new(),
                restore_from: String::new(),
                render: None,
                role: recorded.role.clone(),
                companion_exists: false,
                syntax_invalid: false,
                scope: JournalScope::Agent,
            },
            Vec::new(),
            true,
        ));
    }
    if !safety.reasons.is_empty() {
        if contains_recovery_reason(&safety.reasons, RecoveryReason::NotWritable)
            || contains_recovery_reason(&safety.reasons, RecoveryReason::ParentUnavailable)
        {
            return Err(OperationError::new(
                ErrorCode::ConfigNotWritable,
                format!("{} managed configuration is not writable", agent_name(kind)),
            ));
        }
        return Err(OperationError::new(
            ErrorCode::ConfigInvalid,
            format!("{} managed configuration path is unsafe", agent_name(kind)),
        ));
    }
    let (revision, content, mode) = read_keyed_revision(
        signer,
        Path::new(&recorded.path),
        super::plan::REVISION_CONTEXT_AGENT_FILE,
    )
    .map_err(|_| {
        OperationError::new(
            ErrorCode::ConfigInvalid,
            format!("{} managed configuration cannot be read", agent_name(kind)),
        )
    })?;
    if !revision.exists {
        return Ok((
            PlannedFile {
                agent: Some(kind),
                mode: ConfigMode::Merge,
                format,
                source_path: recorded.path.clone(),
                target_path: recorded.path.clone(),
                operation: Operation::Delete,
                contains_api_key: false,
                preserves: Vec::new(),
                warning: String::new(),
                source_revision: revision.clone(),
                target_revision: revision,
                source_content: Vec::new(),
                source_mode: 0,
                target_mode: 0,
                backup_required: false,
                backup_source: String::new(),
                backup_path: String::new(),
                restore_from: String::new(),
                render: None,
                role: recorded.role.clone(),
                companion_exists: false,
                syntax_invalid: false,
                scope: JournalScope::Agent,
            },
            Vec::new(),
            true,
        ));
    }
    super::preview::ensure_writable(&recorded.path)?;
    let transform = match kind {
        Agent::Claude => {
            let root = decode_object(&content).ok_or_else(|| {
                OperationError::new(
                    ErrorCode::ConfigInvalid,
                    "Claude Code configuration is not a JSON object",
                )
            })?;
            cleanup_claude(&root, &section.owned_paths)?
        }
        Agent::OpenCode => {
            let parsed = if format == Format::Jsonc {
                strip_jsonc(&content).map_err(|_| {
                    OperationError::new(
                        ErrorCode::ConfigInvalid,
                        "opencode configuration is invalid JSONC",
                    )
                })?
            } else {
                content.clone()
            };
            let root = decode_object(&parsed).ok_or_else(|| {
                OperationError::new(
                    ErrorCode::ConfigInvalid,
                    "opencode configuration is not a JSON object",
                )
            })?;
            cleanup_opencode(
                &root,
                section.owned_paths.iter().any(|path| path == "model"),
            )?
        }
        Agent::Codex if recorded.role == "config" => {
            let saved = saved.codex.as_ref().ok_or_else(|| {
                OperationError::new(
                    ErrorCode::ConfigInvalid,
                    "saved Codex configuration is invalid",
                )
            })?;
            cleanup_codex_config(&content, saved)?
        }
        Agent::Codex => {
            let root = decode_object(&content).ok_or_else(|| {
                OperationError::new(
                    ErrorCode::ConfigInvalid,
                    "Codex auth configuration is not a JSON object",
                )
            })?;
            cleanup_codex_auth(&root)?
        }
    };
    let operation = if transform.delete {
        Operation::Delete
    } else {
        Operation::Replace
    };
    let mut file = PlannedFile {
        agent: Some(kind),
        mode: ConfigMode::Merge,
        role: recorded.role.clone(),
        format,
        source_path: recorded.path.clone(),
        target_path: recorded.path.clone(),
        operation,
        contains_api_key: kind != Agent::Codex || recorded.role == "auth",
        preserves: Vec::new(),
        warning: String::new(),
        source_revision: revision.clone(),
        target_revision: revision.clone(),
        source_content: content,
        source_mode: mode,
        target_mode: mode,
        backup_required: true,
        backup_source: recorded.path.clone(),
        backup_path: String::new(),
        restore_from: recorded.path.clone(),
        render: None,
        companion_exists: false,
        syntax_invalid: false,
        scope: JournalScope::Agent,
    };
    if operation == Operation::Replace {
        file.render = Some(FileRender::Static(transform.content));
    }
    Ok((
        file,
        transform.removed_paths,
        revision.token_value() != recorded.revision_mac,
    ))
}

fn cleanup_file_format(kind: Agent, recorded: &LastAppliedFile) -> Format {
    if kind == Agent::Codex && recorded.role == "config" {
        return Format::Toml;
    }
    if kind == Agent::OpenCode
        && Path::new(&recorded.path)
            .extension()
            .and_then(|value| value.to_str())
            == Some("jsonc")
    {
        return Format::Jsonc;
    }
    Format::Json
}

pub fn cleanup_token_for_plan(
    signer: &TokenSigner,
    plan: &CleanupPlan,
) -> Result<String, OperationError> {
    signer
        .sign_cleanup_revision(CleanupRevisionClaims {
            agent: plan.write.selected[0],
            files: plan.token_files.clone(),
            state_operation: plan.state_operation.as_str().to_owned(),
            state_revision: plan.write.sidecar_revision.claim(),
            managed_config_drift: plan.managed_config_drift(),
        })
        .map_err(|_| {
            OperationError::new(
                ErrorCode::ModelStateInvalid,
                "Agent cleanup revision token could not be created",
            )
        })
}

pub fn cleanup_preview_for_plan(
    signer: &TokenSigner,
    state_dir: &Path,
    plan: &CleanupPlan,
) -> Result<CleanupPreview, OperationError> {
    let token = cleanup_token_for_plan(signer, plan)?;
    let sidecar = sidecar_path(state_dir).to_string_lossy().into_owned();
    let mut preview = CleanupPreview {
        revision_token: token,
        agent: plan.write.selected[0],
        files: Vec::new(),
        removed_paths: Vec::new(),
        managed_config_drift: plan.managed_config_drift(),
        state_change: Some(FilePreview {
            role: "state".into(),
            path: sidecar.clone(),
            source_path: String::new(),
            format: Format::Json,
            operation: plan.state_operation,
            operations: vec![plan.state_operation],
            contains_api_key: false,
            preserves: Vec::new(),
            backup: cleanup_backup_plan(&sidecar, false),
            warning: String::new(),
        }),
        state_backup: Some(FilePreview {
            role: "state".into(),
            path: sidecar.clone(),
            source_path: String::new(),
            format: Format::Json,
            operation: Operation::Backup,
            operations: vec![Operation::Backup],
            contains_api_key: false,
            preserves: Vec::new(),
            backup: cleanup_backup_plan(&sidecar, false),
            warning: String::new(),
        }),
    };
    for file in &plan.write.files {
        if file.scope == JournalScope::ManagerState {
            continue;
        }
        let removed = plan
            .removed_by_path
            .get(&file.target_path)
            .cloned()
            .unwrap_or_default();
        preview.removed_paths.extend(removed);
        preview.files.push(FilePreview {
            role: file.role.clone(),
            path: file.target_path.clone(),
            source_path: String::new(),
            format: file.format,
            operation: file.operation,
            operations: vec![file.operation],
            contains_api_key: file.contains_api_key,
            preserves: Vec::new(),
            backup: cleanup_backup_plan(&file.backup_source, true),
            warning: String::new(),
        });
    }
    preview.removed_paths.sort();
    preview.removed_paths.dedup();
    Ok(preview)
}

fn cleanup_backup_plan(path: &str, sensitive: bool) -> BackupPlan {
    BackupPlan {
        required: true,
        sensitive,
        pattern: format!("{path}.bak-<timestamp>-<random>"),
        warning: String::new(),
    }
}

pub fn ensure_cleanup_state_exists(state_dir: &Path, kind: Agent) -> Result<(), OperationError> {
    match std::fs::metadata(sidecar_path(state_dir)) {
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Err(OperationError::new(
            ErrorCode::AgentNotManaged,
            format!("{} is not managed by CodeasierRouter", agent_name(kind)),
        )),
        Err(_) => Err(OperationError::new(
            ErrorCode::ModelStateInvalid,
            "Agent model state cannot be read",
        )),
        Ok(_) => Ok(()),
    }
}

pub fn acquire_cleanup_preview_lock(
    state_dir: &Path,
    deadline: Option<std::time::Instant>,
) -> Result<super::lock::TransactionLock, OperationError> {
    acquire_existing_transaction_lock(state_dir, deadline).map_err(|error| {
        if error.code == ErrorCode::AgentOperationBusy {
            error
        } else {
            OperationError::new(
                ErrorCode::ModelStateInvalid,
                "Agent cleanup coordination state is invalid",
            )
        }
    })
}
